pypeman/plugins/base.py

```diff
@@ -0,0 +1,137 @@
+        """
+        self.name = __name__ + str(self.__class__)
+        logger.debug("Initialized plugin %s", self.name)
+        self.do_init()
+        self.status = self.INITIALIZED
+        self.loop = None
+        self.to_start = []
+        self.start_results = []
+
+    def do_init(self):
+        pass
+
+    def ready(self):
+        """
+        Code to be executed after the whole pypeman graph has been generated.
+        ready() functions of all plugins are executed before the loop is generated
+        """
+        logger.debug("called ready for plugin %s", self.name)
+        assert self.status == self.INITIALIZED
+        self.do_ready()
+        self.status = self.READY
+
+    def do_ready(self):
+        pass
+
+    def set_loop(self, loop):
+        self.loop = loop
+
+    def set_starting(self):
+        logger.debug("starting plugin %s", self.name)
+        assert self.status == self.READY
+        self.status = self.STARTING
+
+    def set_started(self):
+        self.status = self.STARTED
+
+    async def start(self):
+        """
+        Code to be executed after the creation of the event loop
+        """
+        self.set_starting()
+        if iscoroutinefunction(self.do_start):
+            logger.debug("awaiting coro do_start()")
+            future = await self.do_start()
+            logger.debug("started => %s", repr(future))
+            self.start_results.append(future)
+        else:
+            start_rslt = self.do_start()
+            logger.debug("do_start rslt = %s", repr(start_rslt))
+            if isinstance(start_rslt, (list, tuple)):
+                logger.debug("isnocoro")
+                for coro in start_rslt:
+                    rslt = await coro
+                    self.start_results.append(rslt)
+            elif start_rslt is None:
+                pass
+                logger.debug("None")
+            elif iscoroutinefunction(start_rslt):
+                logger.debug("iscoro")
+                rslt = await start_rslt()
+                logger.debug("rslt: %s", repr(rslt))
+                self.start_results.append(rslt)
+            else:
+                logger.debug("awaitable")
+                rslt = await start_rslt
+                logger.debug("rslt: %s", repr(rslt))
+                self.start_results.append(rslt)
+        self.set_started()
+        logger.debug("end of start")
+
+    async def do_start(self):
+        pass
+
+    def start_coro_list(self):
+        """
+        alternative start implementation returning a list of coros to be executed.
+        """
+        return [self.start()]
+
+    async def stop(self):
+        """
+        Code to be executed before mainloop shall be stopped
+        """
+        logger.debug("stopping plugin %s", self.name)
+        assert self.status in (self.INITIALIZED, self.STARTING, self.STARTED)
+        if self.status == self.STARTING:
+            raise NotImplementedError("had to wait till started before stopping")
```

Comment:
NotImplementedError imply you can implement it, but i don't thinks this is the case here. See https://docs.python.org/fr/3.5/library/exceptions.html#NotImplementedError

It's more a [RuntimeError](https://docs.python.org/fr/3.5/library/exceptions.html#RuntimeError) or a pypeman specific exception, don't you think ?

Reply:
can change, though I really wanted to say, that currently it is not implemented to stop any plugin, that has not been started.

In my opinion (but I have to check) it should also be possible to stop a starting plugin, thats why I chose `NotImplementedError`.

But I don't insist on `NotImplemented`.

Reply:
Your opinion?

Reply:
will change to `RuntimeError` if my explanation is not conclusive

Reply:
As you want, I don't insist too ;-)

Reply:
done in ce896e290d55b4be0817dcf3999a94914ef42ce7